tests/test_utils.py

```diff
@@ -496,33 +464,159 @@ def test_prepare_workspace(self):
+            0.0: {
+                'a': 0.0,
+                'b': 'x',
+                'foo': -1.0,
+                'bar': 'bar',
+                '_': 'apple'
+                },
+            }
+        self.assertDictEqual(result, expected_dict)
+    
+    def test_unique_key_not_first_column(self):
+        """utils: test success when key field is not first column."""
+        from natcap.invest import utils
+        csv_text = ("desc,lucode,val1,val2\n"
+                    "corn,1,0.5,2\n"
+                    "bread,2,1,4\n"
+                    "beans,3,0.5,4\n"
+                    "butter,4,9,1")
+        table_path = os.path.join(self.workspace_dir, 'table.csv')
+        with open(table_path, 'w') as table_file:
+            table_file.write(csv_text)
 
-        csv_file = os.path.join(self.workspace, 'csv.csv')
-        with open(csv_file, 'w') as file_obj:
-            file_obj.write(textwrap.dedent(
-                """header1,header2,header3
-                1,2,3
-                4,FOO,bar
-                """
-            ))
+        result = utils.build_lookup_from_csv(
+            table_path, 'lucode', to_lower=True)
+        expected_result = {
+                1: {'desc': 'corn', 'val1': 0.5, 'val2': 2, 'lucode': 1},
+                2: {'desc': 'bread', 'val1': 1, 'val2': 4, 'lucode': 2},
+                3: {'desc': 'beans', 'val1': 0.5, 'val2': 4, 'lucode': 3},
+                4: {'desc': 'butter', 'val1': 9, 'val2': 1, 'lucode': 4}}
+
+        self.assertDictEqual(result, expected_result)
+    
+    def test_non_unique_keys(self):
+        """utils: test error is raised if keys are not unique."""
+        from natcap.invest import utils
+        csv_text = ("lucode,desc,val1,val2\n"
+                    "1,corn,0.5,2\n"
+                    "2,bread,1,4\n"
+                    "2,beans,0.5,4\n"
+                    "4,butter,9,1")
+        table_path = os.path.join(self.workspace_dir, 'table.csv')
+        with open(table_path, 'w') as table_file:
+            table_file.write(csv_text)
 
         with self.assertRaises(ValueError):
-            utils.build_lookup_from_csv(csv_file, 'some_key')
+            utils.build_lookup_from_csv(table_path, 'lucode', to_lower=True)
+    
+    def test_missing_key_field(self):
+        """utils: test error is raised when missing key field."""
+        from natcap.invest import utils
+        csv_text = ("luode,desc,val1,val2\n"
+                    "1,corn,0.5,2\n"
+                    "2,bread,1,4\n"
+                    "3,beans,0.5,4\n"
+                    "4,butter,9,1")
+        table_path = os.path.join(self.workspace_dir, 'table.csv')
+        with open(table_path, 'w') as table_file:
+            table_file.write(csv_text)
+
+        with self.assertRaises(KeyError):
+            utils.build_lookup_from_csv(table_path, 'lucode', to_lower=True)
+    
+    def test_nan_holes(self):
+        """utils: test empty strings returned when missing data is present."""

```

Comment:
Yea, the way it was implemented prior, missing `key_field` was acceptable and I even think a few models take advantage of this. I remember deciding that it made sense to me to skip rows that did not have a value for `key_field`, or to error, but some models tests failed because of it. So I reverted back to just implementing how it was already done.

Reply:
Oh okay. Yeah it looks like the dict ends up with an empty string as a key when that happens. I guess that's fine since that's literally the data that was provided.